app/models/game.rb

```diff
@@ -87,6 +87,108 @@ def is_in_check?(player)
     return false
   end
 
+  def is_in_checkmate?(player)
+    return false if !is_in_check?(player)
+    if player == white_player
+      king = pieces.where(type: "King", player: white_player).first
+      return false if king.can_escape_from_check?
+
+      # Finds the piece that is currently threatening the king
+      king_x = king.position_x
+      king_y = king.position_y
+      threatening_piece = nil
+      black_player.pieces.where(game: self, is_active: true).each do |piece|
+        if piece.type != "King"
+          threatening_piece = piece if piece.is_valid_capture?(king_x, king_y) && piece.is_valid_move?(king_x, king_y)
+        end
+      end
+
+      # Finds the friendly pieces that can block the threatening piece from checking the king
+      white_player.pieces.where(game: self, is_active: true).each do |piece|
+        return false if valid_check_block?(threatening_piece, piece, king) && piece.type != "King"
+      end
+      return true
+    else
+      king = pieces.where(type: "King", player: black_player).first
+      return false if king.can_escape_from_check?
+
+      # Finds the piece that is currently threatening the king
+      king_x = king.position_x
+      king_y = king.position_y
+      threatening_piece = nil
+      white_player.pieces.where(game: self, is_active: true).each do |piece|
+        if piece.type != "King"
+          threatening_piece = piece if piece.is_valid_capture?(king_x, king_y) && piece.is_valid_move?(king_x, king_y)
+        end
+      end
+
+      # Finds the friendly pieces that can block the threatening piece from checking the king
+      black_player.pieces.where(game: self, is_active: true).each do |piece|
+        return false if valid_check_block?(threatening_piece, piece, king) && piece.type != "King"
+      end
+
+      return true
+    end
+  end
+
+  def valid_check_block?(threatening_piece, piece, king)
```

Comment:
This method looks like it might be reducible using some other more generic methods. I just took a quick glance so I can't say much more than that.